clients/src/test/java/org/apache/kafka/common/network/SslVersionsTransportLayerTest.java

```diff
@@ -0,0 +1,133 @@
+ * Checks different versions of the protocol usage on the server and client.
+ */
+@RunWith(value = Parameterized.class)
+public class SslVersionsTransportLayerTest {
+    private static final int BUFFER_SIZE = 4 * 1024;
+    private static final Time TIME = Time.SYSTEM;
+
+    private final List<String> tlsServerProtocols;
+    private final List<String> tlsClientProtocols;
+
+    @Parameterized.Parameters(name = "tlsServerProtocol={0},tlsClientProtocol={1}")
+    public static Collection<Object[]> data() {
+        List<Object[]> values = new ArrayList<>();
+        values.add(new Object[] {Arrays.asList("TLSv1.2"), Arrays.asList("TLSv1.2")});
+        if (Java.IS_JAVA11_COMPATIBLE) {
+            values.add(new Object[] {Arrays.asList("TLSv1.2"), Arrays.asList("TLSv1.3")});
+            values.add(new Object[] {Arrays.asList("TLSv1.3"), Arrays.asList("TLSv1.2")});
+            values.add(new Object[] {Arrays.asList("TLSv1.3"), Arrays.asList("TLSv1.3")});
+            values.add(new Object[] {Arrays.asList("TLSv1.2", "TLSv1.3"), Arrays.asList("TLSv1.3")});
+            values.add(new Object[] {Arrays.asList("TLSv1.2", "TLSv1.3"), Arrays.asList("TLSv1.2")});
+            values.add(new Object[] {Arrays.asList("TLSv1.3"), Arrays.asList("TLSv1.2", "TLSv1.3")});
+            values.add(new Object[] {Arrays.asList("TLSv1.2"), Arrays.asList("TLSv1.2", "TLSv1.3")});
+            values.add(new Object[] {Arrays.asList("TLSv1.2", "TLSv1.3"), Arrays.asList("TLSv1.2", "TLSv1.3")});
+        }
+        return values;
+    }
+
+    public SslVersionsTransportLayerTest(List<String> tlsServerProtocols, List<String> tlsClientProtocols) {
+        this.tlsServerProtocols = tlsServerProtocols;
+        this.tlsClientProtocols = tlsClientProtocols;
+    }
+
+    /**
+     * Tests that connection success with the default TLS version.
+     */
+    @Test
+    public void testTlsDefaults() throws Exception {
+        // Create certificates for use by client and server. Add server cert to client truststore and vice versa.
+        CertStores serverCertStores = new CertStores(true, "server",  "localhost");
+        CertStores clientCertStores = new CertStores(false, "client", "localhost");
+
+        Map<String, Object> sslClientConfigs = getTrustingConfig(clientCertStores, serverCertStores, tlsClientProtocols);
+        Map<String, Object> sslServerConfigs = getTrustingConfig(serverCertStores, clientCertStores, tlsServerProtocols);
+
+        NioEchoServer server = NetworkTestUtils.createEchoServer(ListenerName.forSecurityProtocol(SecurityProtocol.SSL),
+             SecurityProtocol.SSL,
+             new TestSecurityConfig(sslServerConfigs),
+             null,
+            TIME);
+        Selector selector = createSelector(sslClientConfigs);
+
+        String node = "0";
+        selector.connect(node, new InetSocketAddress("localhost", server.port()), BUFFER_SIZE, BUFFER_SIZE);
+
+        if (tlsServerProtocols.contains(tlsClientProtocols.get(0))) {
```

Comment:
Shouldn't this check if the intersection of server and client protocols is non empty?

Reply:
Fixed.